src/lib/FreeC/Pass/KindCheckPass.hs

```diff
@@ -0,0 +1,180 @@
+--   applications. In these cases, a fatal error is reported. Otherwise, the
+--   pass finishes with the given module unchanged.
+--
+--   == Postcondition
+--
+--   All type constructors in type expressions are applied to the correct
+--   number of arguments and there are no type variables or function types on
+--   the left-hand side of type applications.
+module FreeC.Pass.KindCheckPass
+  ( kindCheckPass
+  )
+where
+
+import           Control.Monad                  ( when )
+import           Data.Maybe                     ( fromMaybe )
+
+import           FreeC.Environment              ( lookupArity )
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.Monad.Converter
+import           FreeC.Monad.Reporter
+import           FreeC.Pass
+import           FreeC.Pretty                   ( showPretty )
+
+-- | A compiler pass that checks whether there are type applications with type
+--   variables or function types on the left-hand side in the module and checks
+--   whether all type construcors in the module are fully applied.
+kindCheckPass :: Pass IR.Module
+kindCheckPass m@(IR.Module _ _ _ typeDecls typeSigs _ funcDecls) = do
+  mapM_ checkTypeDecl typeDecls
+  mapM_ checkTypeSig  typeSigs
+  mapM_ checkFuncDecl funcDecls
+  return m
+
+-- | Checks whether all type expressions in a type declaration are correct
+checkTypeDecl :: IR.TypeDecl -> Converter ()
+checkTypeDecl (IR.DataDecl    _ _ _ conDecls) = mapM_ checkConDecl conDecls
+checkTypeDecl (IR.TypeSynDecl _ _ _ typ     ) = checkType typ
+
+-- | Checks whether the arguments of a constructor declaration are kind-correct
+checkConDecl :: IR.ConDecl -> Converter ()
+checkConDecl (IR.ConDecl _ _ types) = mapM_ checkType types
+
+-- | Checks whether the type schema in a type signature is kind-correct
+checkTypeSig :: IR.TypeSig -> Converter ()
+checkTypeSig (IR.TypeSig _ _ typeSchema) = checkTypeSchema typeSchema
+
+-- | Checks whether a given type schema is kind-correct
+checkTypeSchema :: IR.TypeSchema -> Converter ()
+checkTypeSchema (IR.TypeSchema _ _ typ) = checkType typ
+
+-- | Checks whether a function declaration has kind-correct type annotations
+checkFuncDecl :: IR.FuncDecl -> Converter ()
+checkFuncDecl (IR.FuncDecl _ _ _ varPats retType rhs) = do
+  mapM_ checkVarPat varPats
+  mapM_ checkType   retType
+  checkExpr rhs
+  return ()
+
+-- | Checks whether all type annotations in an expression are kind-correct
+checkExpr :: IR.Expr -> Converter ()
+checkExpr (IR.Con _ _ typeSchema      ) = mapM_ checkTypeSchema typeSchema
+checkExpr (IR.Var _ _ typeSchema      ) = mapM_ checkTypeSchema typeSchema
+checkExpr (IR.App _ lhs rhs typeSchema) = do
+  checkExpr lhs
+  checkExpr rhs
+  mapM_ checkTypeSchema typeSchema
+checkExpr (IR.TypeAppExpr _ lhs rhs typeSchema) = do
+  checkExpr lhs
+  checkType rhs
+  mapM_ checkTypeSchema typeSchema
+checkExpr (IR.If _ cond thenExp elseExp typeSchema) = do
```

Comment:
Try to abbreviate consistently. Expression are abbreviated `expr` instead of `exp` in `IR`.

```suggestion
checkExpr (IR.If _ cond thenExpr elseExpr typeSchema) = do
```

In some other places you wrote `typ` to abbreviate types. Unfortunately, the identifier `type` cannot be used. Thus, `type` is "abbreviated" `typeExpr` by most compiler passes.